### PR TITLE
multiple fixes for improved lifecycle and errors

### PR DESCRIPTION
**Pull request by JohnPhoto**
* Init and _internalInit is actually awaitable
* this.bleDevices is reset on init
* check for actual data in write to preven length of undefined
* Throttled re-inits on write-errors.

This PR is a suggestion for general improvement.

My setup:
5 Plejd-devices and multiple scenes which was very much affected by the https://github.com/icanos/hassio-plejd/issues/134, resulting in weird behaviour where some scene-switches worked halfway and some lights was updateable sometimes in HA.
Basically write was given broken data and reinits was buggy resulting in failed retries.

With this PR both the error (but not the rootcause) is prevented and also general improvements to the lifecycles if other errors would occur.

If this PR is approved it will make the following PR's redundant:
* https://github.com/icanos/hassio-plejd/pull/135
* https://github.com/icanos/hassio-plejd/pull/136

### CODE REVIEW

File: plejd/ble.bluez.js [hunk 1]
```diff
@@ -146,10 +147,7 @@ class PlejdService extends EventEmitter {
       console.log('plejd-ble: error: failed to start discovery. Make sure no other add-on is currently scanning.');
       return;
     }
-
-    setTimeout(async () => {
-      await this._internalInit();
-    }, this.connectionTimeout * 1000);
+    return new Promise(resolve => setTimeout(() => resolve(this._internalInit()), this.connectionTimeout * 1000));
```
Comment (JohnPhoto) [inline plejd/ble.bluez.js:150]: This allows us to actually wait for the `_internalInit` in `write` errorhandler so we do not retry into void.

File: plejd/ble.bluez.js [hunk 2]
```diff
@@ -85,8 +85,9 @@ class PlejdService extends EventEmitter {
       this.objectManager.removeAllListeners();
     }
 
+    this.bleDevices = [];
```
Comment (JohnPhoto) [inline plejd/ble.bluez.js:88]: We need to reset this to not continue the list on a re-init (write/ping errorhandler)

File: plejd/ble.bluez.js [hunk 3]
```diff
@@ -379,8 +377,20 @@ class PlejdService extends EventEmitter {
     this.characteristics.lastData.StartNotify();
   }
 
+  async throttledInit(delay) {
```
Comment (JohnPhoto) [inline plejd/ble.bluez.js:380]: If multiple errors is thrown before the timeout is done, we do not need to re-init multiple times (which results is even more errors since we teared everything down when trying to re-init while re-initing)

File: plejd/ble.bluez.js [hunk 4]
```diff
@@ -379,8 +377,20 @@ class PlejdService extends EventEmitter {
...
+     const result = await this.init();
+     this.delayedInit = null;
+     resolve(result)
+   }, delay))
+   return this.delayedInit;
+  }
+
   async write(data, retry = true) {
-    if (!this.plejdService || !this.characteristics.data) {
+    if (!data || !this.plejdService || !this.characteristics.data) {
```
Comment (JohnPhoto) [inline plejd/ble.bluez.js:393]: Apparently it is not unusual to get `!data` atleast with multiple Plejddevices and scenes.

File: plejd/ble.bluez.js [hunk 5]
```diff
@@ -393,16 +403,13 @@ class PlejdService extends EventEmitter {
         setTimeout(() => this.write(data, retry), 1000);
         return;
       }
-
       console.log('plejd-ble: write failed ' + err);
-      setTimeout(async () => {
-        await this.init();
+      await this.throttledInit(this.connectionTimeout * 1000);
```
Comment (JohnPhoto) [inline plejd/ble.bluez.js:407]: hmmm.. I just realized that this actually changed the entire write-flow. Since this might actually stop any writes at all until we have re-init'ed.
On the flip side: what sense does it make to try to write to a channel we know is broken?

File: plejd/ble.bluez.js [hunk 6]
```diff
@@ -379,8 +377,20 @@ class PlejdService extends EventEmitter {
     this.characteristics.lastData.StartNotify();
   }
 
+  async throttledInit(delay) {
+    if(this.delayedInit){
```
Comment (SweVictor) [inline plejd/ble.bluez.js:381]: Code style mainly, but may I suggest initializing `this.delayedInit` (and possibly other new class variables) in the constructor?

**Comment by icanos** [comment]
Thanks! I'll review this asap and get back to you :-)

File: plejd/ble.bluez.js (hunk 5 above)
Comment (SweVictor) [reply plejd/ble.bluez.js:407]: Agree, this looks suspicious. Since JS is single-threaded it's probably ok, but I would feel safer if we just put back the message in the write queue and spin of a new `init` without `await` (let the `write` function return)

Comment (JohnPhoto) [reply plejd/ble.bluez.js:407]: Yes, but it is quite important to keep the inits throttled / ”singletoned” because the current setup result in dbus errors presented by in #104

Comment (SweVictor) [reply plejd/ble.bluez.js:407]: True, but I'm thinking row 381 solves that?

File: plejd/ble.bluez.js (hunk 6 above)
Comment (JohnPhoto) [reply plejd/ble.bluez.js:381]: Absolutely, I just tried stuff out at 2:30 in the night!

**Comment by kodarn** [comment]
@JohnPhoto Great patch!

File: plejd/ble.bluez.js (hunk 5 above)
Comment (JohnPhoto) [reply plejd/ble.bluez.js:408]: Will experiment with putting it back on the writeQ now and just let it return as your suggested.

Comment (SweVictor) [reply plejd/ble.bluez.js:408]: Nice. Maybe we mean the same thing, but just so I'm clear: I merely mean removing the "awaiting" on throttledInit, not remove the call completely.

Comment (JohnPhoto) [reply plejd/ble.bluez.js:408]: I know 👍 
The problem I am facing with that though.. is that for each "round" of runWriteQueue, we try to empty the queue before we allow us to timeout.. basically:
```
while q.length > 1
   if(error) q.push(data)
// => everlasting loop
```

So either write needs to flag fail / throw to be able to break the bad loop.. or keep current

Did find a few clearInterval-instead-of-clearTimeout bugs though which I will commit regardless of this convo.

Comment (SweVictor) [reply plejd/ble.bluez.js:408]: Ok, I see I actually have a new strategy for this (per-device queue, better retry strategy etc) in the works, so maybe it would be good to stop "as-is" with this PR and then see where we are. Main thing to me is that each PR doesn't degrade existing functionality. We don't need to solve all problems at once ;)

Comment (JohnPhoto) [reply plejd/ble.bluez.js:408]: This is my final atempt in that case. Basically rewriting who is responsible for inits:
https://github.com/JohnPhoto/hassio-plejd/pull/5

Feel free to take a look and see which one you prefer (That PR is targeting this PR to highlight the diffs agains work already done here)

Comment (SweVictor) [reply plejd/ble.bluez.js:408]: I prefer this one compared to the JohnPhoto#5 PR. I think `retryWriteQueue` is a nice idea but solves the problem possibly in the wrong way. I have experimented quite a bit with transitions, retries, and so on and come to the conclusion that you typically only want to send the latest command to each queue (if you set brightness 1 => 50 => 255 you don't want 1 to be retried).

So I suggest merging this and I'll see if I can fix the bug I have when transitioning to off (light sometimes step up brightness before dimming down). WIP is here if you want to look at the idea. It currently implements "shouldRetry" on each command, but could easily be extended to a retryCount > 1. Not rewritten after improvements in this PR obviously :)

Comment (JohnPhoto) [reply plejd/ble.bluez.js:408]: Fair enough. I just started with this yesterday so I’m new here 😀

Comment (SweVictor) [reply plejd/ble.bluez.js:408]: I beat you to this project with a few days then 😉

**Comment by icanos** [comment]
Thanks for a great PR, I’ll merge this into the project.

**Comment by SweVictor** [comment]
Closed issue but still - I found the "real real" reason why we get UnhandledPromiseRejectionWarnings... It's because the `dbus-next` `systemBus` is implemented as an `eventEmitter` that will emit `error` events AND throw errors whenever an error occurs. 

And by NodeJS conventions, emitted `error` events will log angry messages and soon crash the process... 

Addressed in https://github.com/SweVictor/hassio-plejd/commit/dca491bf009ecd229288bc98634501bcd374e6bd
